File: src/utilities.ts
import moment, { Moment } from 'moment'
import Person from './interfaces/Person'

export interface Sorting {
  column: string
  direction: 'asc' | 'desc'
}

export function sortPersons(persons: Person[], sorting: Sorting): Person[] {
  // Init list of ordered persons
  let orderedPersons: Person[] = []

  // If persons are present
  if (persons) {
    try {
      if (sorting.column === 'dateOfBirthWithoutYear') {
        orderedPersons = [...persons].sort((a: Person, b: Person) => {
          if (
            sorting.direction === 'asc'
              ? a.dateOfBirth.substring(5) > b.dateOfBirth.substring(5)
              : a.dateOfBirth.substring(5) < b.dateOfBirth.substring(5)
          )
            return 1
          return -1
        })
      } else {
        // Order list of persons by string column
        orderedPersons = [...persons].sort((a: Person, b: Person) => {
          const valueA = (
            a[sorting.column as keyof typeof a] as string
          ).toLowerCase()
          const valueB = (
            b[sorting.column as keyof typeof b] as string
          ).toLowerCase()
          if (sorting.direction === 'asc' ? valueA > valueB : valueA < valueB)
            return 1
          return -1
        })
      }
    } catch {
      // Order list of persons by number column
      orderedPersons = [...persons].sort((a: Person, b: Person) => {
        const valueA = a[sorting.column as keyof typeof a] as number
        const valueB = b[sorting.column as keyof typeof b] as number
        if (sorting.direction === 'asc' ? valueA > valueB : valueA < valueB)
          return 1
        return -1
      })
    } finally {
      return orderedPersons
    }
  }
  return orderedPersons
}

export function getAnniversariesOfDateInTimespan(
  date: Moment,
  timespanStart: Moment,
  timespanEnd: Moment
): number[] {
  // Copy parameters to local instances
  date = moment(date)
  timespanStart = moment(timespanStart)
  timespanEnd = moment(timespanEnd)

  // If end is before start
  if (timespanEnd.isBefore(timespanStart)) {
    // Swap start and end
    let temp = timespanEnd
    timespanEnd = timespanStart
    timespanStart = temp
  }

  // Init list for anniversaries
  let anniversaries = []

  // Init list of tuples for separate timespans
  let timespans = []

  // Divide timespan in parts for separate years
  while (timespanEnd.year() > timespanStart.year()) {
    let subTimespanStart = moment(timespanStart)
    let subTimespanEnd = moment(timespanStart.endOf('year'))
    timespans.push([subTimespanStart, subTimespanEnd])

    timespanStart = timespanStart.add(1, 'year').startOf('year')
  }
  timespans.push([timespanStart, timespanEnd.endOf('day')])

  // For each timespan
  for (let ts of timespans) {
    let year = ts[0].year()
    let dateInYear = moment(date).year(year)

    // If date is in timespan (include start and end)
    if (dateInYear.isBetween(ts[0], ts[1], undefined, '[]')) {
      anniversaries.push(year - date.year())
    }
  }

  return anniversaries
}

File: src/MemberManagement/Client/Shared/Utilities.cs
using Microsoft.AspNetCore.Components;
using System.Web;

namespace MemberManagement.Client.Shared
{
    public class Utilities
    {
        private NavigationManager NavigationManager { get; set; } = default!;

        public Utilities(NavigationManager navigationManager)
        {
            NavigationManager = navigationManager;
        }

        public static List<int> GetAnniversariesOfDateInTimespan(DateTime date, DateTime timespanStart, DateTime timespanEnd)
        {
            // If end is before start
            if (timespanEnd < timespanStart)
            {
                // Swap start and end
                (timespanStart, timespanEnd) = (timespanEnd, timespanStart);
            }

            List<int> anniversaries = new();

            // Init list of tuples for separate timespans
            List<Tuple<DateTime, DateTime>> timespans = new();

            // Divide timespan in parts for separate years
            while (timespanEnd.Year > timespanStart.Year)
            {
                DateTime subTimespanStart = timespanStart;
                DateTime subTimespanEnd = new(timespanStart.Year, 12, 31, 23, 59, 59);
                timespans.Add(new(subTimespanStart, subTimespanEnd));

                timespanStart = new(timespanStart.Year + 1, 1, 1);
            }
            timespans.Add(new(timespanStart, new(timespanEnd.Year, timespanEnd.Month, timespanEnd.Day, 23, 59, 59)));

            PrintTimespans(timespans);

            // For each timespan
            foreach (var timespan in timespans)
            {
                var year = timespan.Item1.Year;
                DateTime dateInYear = new(year, date.Month, date.Day);
                if (timespan.Item1 <= dateInYear && dateInYear <= timespan.Item2)
                {
                    anniversaries.Add(year - date.Year);
                }
            }

            return anniversaries;
        }

        public string GetQueryStringParameter(string parameterName)
        {
            var queryVariables = HttpUtility.ParseQueryString(new Uri(NavigationManager.Uri).Query);
            if (queryVariables.AllKeys.Contains(parameterName))
            {
                var origin = queryVariables[parameterName] ?? string.Empty;
                return origin;
            }
            else
            {
                return string.Empty;
            }
        }

        public void NavigateTo(string target, bool preserveOrigin = true)
        {

            var originPart = "";

            if (preserveOrigin)
            {
                var origin = GetQueryStringParameter("origin");
                if (origin != string.Empty)
                {
                    originPart = $"?origin={origin}";
                }
            }

            NavigationManager.NavigateTo(target + originPart, false);
        }

        private static void PrintTimespans(List<Tuple<DateTime, DateTime>> timespans)
        {
            foreach (var timespan in timespans)
            {
                Console.WriteLine($"{timespan.Item1} - {timespan.Item2}");
            }
        }
    }
}
